compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/DecomposeLinalgExtPackUnPackToAIR.cpp

```diff
@@ -0,0 +1,363 @@
+  SmallVector<int64_t> innerPositionsPerm = computePermutationVector(
+      packedRank, lastDims, packingMetadata.insertPositions);
+
+  SmallVector<int64_t> outerPos = packingMetadata.outerPositions;
+  if (!outerDimsPerm.empty()) applyPermutationToVector(outerPos, outerDimsPerm);
+  SmallVector<int64_t> outerPositionPerm = computePermutationVector(
+      packedRank, packingMetadata.outerPositions, outerPos);
+
+  SmallVector<int64_t> packedToStripMinedShapePerm = innerPositionsPerm;
+  applyPermutationToVector(packedToStripMinedShapePerm, outerPositionPerm);
+
+  return packedToStripMinedShapePerm;
+}
+
+struct LowerPackUnPackResult {
+  memref::TransposeOp transposeOp;
+  xilinx::air::DmaMemcpyNdOp dmaOp;
+};
+
+FailureOr<LowerPackUnPackResult> lowerPack(RewriterBase &rewriter,
+                                           IREE::LinalgExt::PackOp packOp) {
+  // 1. Filter out NYI cases.
+  auto packedMemrefType = packOp.getOutputType();
+  if (llvm::any_of(packOp.getStaticInnerTiles(),
+                   [](int64_t size) { return ShapedType::isDynamic(size); })) {
+    return rewriter.notifyMatchFailure(
+        packOp,
+        "non-static shape NYI, needs a more powerful memref.expand_shape op");
+  }
+
+  Location loc = packOp->getLoc();
+  OpBuilder::InsertionGuard g(rewriter);
+  rewriter.setInsertionPoint(packOp);
+
+  auto innerDimsPos = packOp.getInnerDimsPos();
+  auto destShape = packOp.getOutputType().getShape();
+  SmallVector<int64_t> transpPerm = {};
+  Value tile = nullptr;
+  if (llvm::any_of(innerDimsPos, [destShape](int64_t index) {
+        return destShape[index] != 1;
+      })) {
+    // 1. Computes the permutation vector to shuffle packed shape into the shape
+    // before any outer or inner permutations have been applied.
+    PackingMetadata packingMetadata =
+        computePackingMetadata(packedMemrefType.getRank(), innerDimsPos);
+
+    SmallVector<int64_t> packedToStripMinedShapePerm =
+        getPackUnpackStripMinedPerm(packedMemrefType.getShape(), innerDimsPos,
+                                    packOp.getOuterDimsPerm());
+
+    // 2. Compute the stripMinedShape: this is the packed shape before any outer
+    // or inner permutations have been applied.
+    SmallVector<int64_t> stripMinedShape(packedMemrefType.getShape());
+    applyPermutationToVector(stripMinedShape, packedToStripMinedShapePerm);
+
+    // 3. Expand from the padded result to the stripMinedShape.
+    tile = rewriter.create<memref::ExpandShapeOp>(
+        loc, stripMinedShape, packOp.getInput(),
+        packingMetadata.reassociations);
+
+    // 4. Transpose stripMinedShape to packedShape.
+    transpPerm = invertPermutationVector(packedToStripMinedShapePerm);
+  } else {
+    tile = packOp.getInput();
+    // Transpose the tile to match the inner tile order.
+    auto inputShape = packOp.getInputType().getShape();
+    transpPerm = getPackUnpackRankReducedPerm(inputShape, innerDimsPos,
```

Comment:
Instead of doing this, isnt it easier to just do the `if` part for both the rank-reduced and non-rank-reduced part, and for the rank-reduced case (where all the outer packed dimensions are 1) just drop those dimensions in the permutation vector and then normalize it....

Reply:
Agreed. We could just normalize the `innerDimPos` and use that as permutation vector. Updated.